src/node_file.cc

```diff
@@ -2480,6 +2480,80 @@ static void WriteString(const FunctionCallbackInfo<Value>& args) {
   }
 }
 
+static void WriteFileUtf8(const FunctionCallbackInfo<Value>& args) {
+  // Fast C++ path for fs.writeFileSync(path, data) with utf8 encoding
+  // (file, data, options.flag, options.mode)
+
+  Environment* env = Environment::GetCurrent(args);
+  auto isolate = env->isolate();
+
+  CHECK_GE(args.Length(), 4);
+
+  BufferValue value(isolate, args[1]);
+  CHECK_NOT_NULL(*value);
+
+  CHECK(args[2]->IsInt32());
+  const int flags = args[2].As<Int32>()->Value();
+
+  CHECK(args[3]->IsInt32());
+  const int mode = args[3].As<Int32>()->Value();
+
+  uv_file file;
+  uv_fs_t req;
+
+  bool is_fd = args[0]->IsInt32();
+
+  // Check for file descriptor
+  if (is_fd) {
+    file = args[0].As<Int32>()->Value();
+  } else {
+    BufferValue path(isolate, args[0]);
+    CHECK_NOT_NULL(*path);
+    if (CheckOpenPermissions(env, path, flags).IsNothing()) return;
+
+    FS_SYNC_TRACE_BEGIN(open);
+    file = uv_fs_open(nullptr, &req, *path, flags, mode, nullptr);
+    FS_SYNC_TRACE_END(open);
+
+    if (req.result < 0) {
+      uv_fs_req_cleanup(&req);
+      return env->ThrowUVException(req.result, "open", nullptr, path.out());
+    }
+  }
+
+  uint32_t offset = 0;
+  uint32_t length = value.length();
+
+  uv_buf_t uvbuf = uv_buf_init(*value, length);
+
+  FS_SYNC_TRACE_BEGIN(write);
+  while (offset != length) {
+    int bytesWritten = uv_fs_write(nullptr, &req, file, &uvbuf, 1, -1, nullptr);
+    offset += bytesWritten;
+
+    // Errored out (-1)
+    if (bytesWritten < 0) {
+      FS_SYNC_TRACE_END(write);
+
+      if (!is_fd) {
+        FS_SYNC_TRACE_BEGIN(close);
+        CHECK_EQ(0, uv_fs_close(nullptr, &req, file, nullptr));
+        FS_SYNC_TRACE_END(close);
+      }
+      uv_fs_req_cleanup(&req);
+
+      return env->ThrowUVException(bytesWritten, "write");
+    }
+  }
+  FS_SYNC_TRACE_END(write);
+
+  if (!is_fd) {
+    FS_SYNC_TRACE_BEGIN(close);
+    CHECK_EQ(0, uv_fs_close(nullptr, &req, file, nullptr));
+    FS_SYNC_TRACE_END(close);
+  }
+  uv_fs_req_cleanup(&req);
```

Comment:
I think this could be simplified to something like this (after you make the buffer update change)
```suggestion
  int bytesWritten = 0;
  uint32_t offset = 0;
  uint32_t length = value.length();

  uv_buf_t uvbuf = uv_buf_init(*value, length);

  FS_SYNC_TRACE_BEGIN(write);
  while (offset != length) {
    bytesWritten = uv_fs_write(nullptr, &req, file, &uvbuf, 1, -1, nullptr);
    if (bytesWritten < 0) {
      break;
    }
    offset += bytesWritten;
  }
  FS_SYNC_TRACE_END(write);

  if (!is_fd) {
    FS_SYNC_TRACE_BEGIN(close);
    CHECK_EQ(0, uv_fs_close(nullptr, &req, file, nullptr));
    FS_SYNC_TRACE_END(close);
  }
  uv_fs_req_cleanup(&req);

  if (bytesWritten < 0) {
    return env->ThrowUVException(bytesWritten, "write");
  }
```

Reply:
This is great, thanks!